asr1k_neutron_l3/models/neutron/l3/router.py

```diff
@@ -150,17 +150,22 @@ def get_internal_cidrs(self):
 
     def _build_routes(self):
         routes = route.RouteCollection(self.router_id)
-        primary_route = self._primary_route()
-        if primary_route is not None and self._route_has_connected_interface(primary_route):
-            routes.append(primary_route)
 
+        # In case the customer sets a default route, we will restrain from programming the openstack primary route
+        primary_overridden = False
         for l3_route in self.router_info.get('routes', []):
             ip, netmask = utils.from_cidr(l3_route.get('destination'))
+            if netmask == '0.0.0.0':
+                primary_overridden = True
 
             r = route.Route(self.router_id, ip, netmask, l3_route.get('nexthop'))
             if self._route_has_connected_interface(r):
                 routes.append(r)
 
+        primary_route = self._primary_route()
+        if not primary_overridden and primary_route is not None and self._route_has_connected_interface(primary_route):
+            routes.append(primary_route)
```

Comment:
I know this is a really unfun question, but... You're shifting the position of the `primary_route` in the list of routes for a VRF here. Semantically in the router config this should not make a difference, though when I look at our asr1k configs, the default route is always first in the list. Would this cause a reconfiguration of the list of routes in all VRFs that have more than a single default route?

Option 1: Live with that, it will happen once and probably not cause any problems
Option 2: Test what this actually does on device side, so we know what happens and deal with the result
Option 3: Use `routes.insert(0, primary_route)` to preserve the old order, whether we need it or not.

Reply:
Let's try this out:
CLI config:
```
router#show run | in ip route vrf [redacted]
ip route vrf [redacted] 0.0.0.0 0.0.0.0 [redacted]
ip route vrf [redacted] 100.100.0.0 255.255.255.0 10.180.1.45
ip route vrf [redacted] 100.100.1.0 255.255.255.0 10.180.1.172
ip route vrf [redacted] 100.100.2.0 255.255.255.0 10.180.1.245
ip route vrf [redacted] 100.100.3.0 255.255.255.0 10.180.1.235
router#
```

YANG state of mind:
```
netconf-console --host router --get-config -N 'native=http://cisco.com/ns/yang/Cisco-IOS-XE-native' -x '/native/ip/route/vrf[name="[redacted]"]/ip-route-interface-forwarding-list'
<?xml version='1.0' encoding='UTF-8'?>
<data xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0">
  <native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native">
    <ip>
      <route>
        <vrf>
          <name>[redacted]</name>
          <ip-route-interface-forwarding-list>
            <prefix>0.0.0.0</prefix>
            <mask>0.0.0.0</mask>
            <fwd-list>
              <fwd>[redacted]</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.0.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.45</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.1.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.172</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.2.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.245</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.3.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.235</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
        </vrf>
      </route>
    </ip>
  </native>
</data>
```

Let's alter it:
```
~/devel ❯ cat edit-config.xml                                                           
  <native xmlns="http://cisco.com/ns/yang/Cisco-IOS-XE-native">
    <ip>
      <route>
        <vrf operation="merge">
          <name>[redacted]</name>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.0.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.45</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.1.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.172</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.2.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.245</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>100.100.3.0</prefix>
            <mask>255.255.255.0</mask>
            <fwd-list>
              <fwd>10.180.1.235</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
          <ip-route-interface-forwarding-list>
            <prefix>0.0.0.0</prefix>
            <mask>0.0.0.0</mask>
            <fwd-list>
              <fwd>[redacted]</fwd>
            </fwd-list>
          </ip-route-interface-forwarding-list>
        </vrf>
      </route>
    </ip>
  </native>%

~/devel ❯ netconf-console --host router  --edit-config edit-config.xml
<?xml version='1.0' encoding='UTF-8'?>
<ok xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" xmlns:nc="urn:ietf:params:xml:ns:netconf:base:1.0"/>
```

CLI again:
```
router#show run | in ip route vrf [redacted]
ip route vrf [redacted] 0.0.0.0 0.0.0.0 [redacted]
ip route vrf [redacted] 100.100.0.0 255.255.255.0 10.180.1.45
ip route vrf [redacted] 100.100.1.0 255.255.255.0 10.180.1.172
ip route vrf [redacted] 100.100.2.0 255.255.255.0 10.180.1.245
ip route vrf [redacted] 100.100.3.0 255.255.255.0 10.180.1.235
router#
```

-> Device does not care, nothing in the catchall log.

Now that I checked that I also think, the agent alter the sort-order either, cause we sort the routes by prefix when creating the YANG-dict-thing [here](https://github.com/sapcc/asr1k-neutron-l3/blob/stable/ussuri-m3/asr1k_neutron_l3/models/netconf_yang/route.py#L126).

Reply:
What a nice thing that we sort them, the commit message also explicitly mentions that :)